[clangd] forward clang-tidy's readability-identifier-naming fix to textDocument/rename

tom-anders opened the pull request.

tom-anders committed:
[clangd] forward clang-tidy's readability-identifier-naming fix to textDocument/rename

HighCommander4 commented:
Linking to the issue this is seeking to address for reference: https://github.com/clangd/clangd/issues/1589

HighCommander4 reviewed (commented):
Thanks, the approach in this patch looks pretty good to me.

My only feedback is to encapsulate the "hard coding" into a function like:

```
Option<CodeActionResult::Rename> TryConvertToRename(const Diag *D, const Fix *F)
```

because I can imagine in the future coming across other diagnostics that are conceptually renames that we may want to handle this way.

Regarding testing: I find lit-tests pretty hard to read and maintain; a nicer alternative might be a gtest in `ClangdLSPServerTests` (e.g. see [this one](https://searchfox.org/llvm/rev/23b233c8adad5b81e185e50d04356fab64c2f870/clang-tools-extra/clangd/unittests/ClangdLSPServerTests.cpp#198) for testing call hierarchy incoming calls).

tom-anders commented:
Duplicate issue: https://github.com/clangd/clangd/issues/741

HighCommander4 commented:
> Duplicate issue: [clangd/clangd#741](https://github.com/clangd/clangd/issues/741)

Nice find, I closed it as a duplicate.

tom-anders commented:
Thanks for the feedback!

> Thanks, the approach in this patch looks pretty good to me.
> 
> My only feedback is to encapsulate the "hard coding" into a function like:
> 
> ```
> Option<CodeActionResult::Rename> TryConvertToRename(const Diag *D, const Fix *F)
> ```
> 
> because I can imagine in the future coming across other diagnostics that are conceptually renames that we may want to handle this way.

:heavy_check_mark: 

> Regarding testing: I find lit-tests pretty hard to read and maintain; a nicer alternative might be a gtest in `ClangdLSPServerTests` (e.g. see [this one](https://searchfox.org/llvm/rev/23b233c8adad5b81e185e50d04356fab64c2f870/clang-tools-extra/clangd/unittests/ClangdLSPServerTests.cpp#198) for testing call hierarchy incoming calls).

Perfect, thanks for the hint! I had to adapt `LSPClient` a bit, it usually would flag every server->client call as an error, but in this case we actually expect a `workspace/applyEdit` after we trigger the renaming.

HighCommander4 commented:
Looks pretty good!

Are there further changes you're planning to make, or is this ready to graduate from "Draft" status?

tom-anders commented:
> Looks pretty good!
> 
> Are there further changes you're planning to make, or is this ready to graduate from "Draft" status?

Ah, thanks about the heads up, forgot about this

tom-anders commented:
(fixed formatting issues)

HighCommander4 reviewed (changes requested):
Ok, I had a more detailed look at the implementation. Still looks good overall, just have some minor comments.

HighCommander4 commented:
Something else I noticed while trying out the patch locally: before the patch, the description of the quick fix in the editor is "change 'foo' to 'Foo'", i.e. a description of what the code action will do.

After the patch, the description of the quick fix is now "invalid case style for function 'foo'", i.e. it describes the problem, not the fix.

We should keep the description of the code action the same, i.e. describing the fix.

HighCommander4 commented:
(It's also no longer labelled as a "quick fix", likely due to the `CodeActionKind` changing from `quickfix` to `refactor`, but that's probably fine.)

tom-anders commented on clang-tools-extra/clangd/ClangdServer.cpp:
```diff
@@ -648,6 +649,27 @@ tweakSelection(const Range &Sel, const InputsAndAST &AST,
+// rename commands, such that we can leverage the index for more accurate
+// results.
+std::optional<ClangdServer::CodeActionResult::Rename>
+TryConvertToRename(const Diag *Diag, const ClangdServer::DiagRef &DR,
```
hah :D 

At work we're using Google style, so functions are CamelCase and variables snake_case. Switching back and forth between Google and LLVM style is so confusing

tom-anders commented:
> Something else I noticed while trying out the patch locally: before the patch, the description of the code action in the editor is "change 'foo' to 'Foo'", i.e. a description of what the code action will do.
> 
> After the patch, the description of the code action is now "invalid case style for function 'foo'", i.e. it describes the problem, not the fix.
> 
> We should keep the description of the code action the same, i.e. describing the fix.

Ah good find, I added a new `FixMessage` field for this (and also added this to the test)

HighCommander4 reviewed (approved):
Thanks, LGTM!

HighCommander4 committed:
Merge branch 'main' into clangTidyRename
